crates/re_log_types/src/data_row.rs

```diff
@@ -266,6 +266,30 @@ pub struct DataRow {
 }
 
 impl DataRow {
+    /// Builds a new `DataRow` from anything implementing [`AsComponents`].
+    pub fn from_component_batches(
+        row_id: RowId,
+        timepoint: TimePoint,
+        entity_path: EntityPath,
+        as_components: &dyn AsComponents,
+    ) -> anyhow::Result<Self> {
```

Comment:
How would we get the `num_instances` in that case? Take the max of all the batches? What if there are no batches, or if they are all splats?

Reply:
> Take the max of all the batches?

Yes, that matches the behavior of our log methods (in all languages, even!):
- https://github.com/rerun-io/rerun/blob/main/crates/re_sdk/src/recording_stream.rs#L710-L720
- https://github.com/rerun-io/rerun/blob/main/rerun_py/rerun_sdk/rerun/_log.py#L206-L207

> What if there are no batches

Then there's nothing in the row and there are no instances

> or if they are all splats?

You cannot have "all splats", that would just result in a row with num_instances = 1.

Reply:
Oh, I thought we stored `num_instances` separately.

So what happens if I log a full point cloud and then what to update all the colors with a splat color - that would be a row with `num_instance = 1` then, even though will affect much more instances than that?

Reply:
That's where it becomes funky...

If you do this:
```python
rr.log("random", rr.Points3D(positions, colors=colors, radii=radii))
rr.log_components("random", [rr.components.ColorBatch([255, 0, 0])])
```

Then you're going to end up with the original `colors` being discarded, a single red point and the rest of the points using the default color for this entity path (because that `ColorBatch` is _not_ a splat).

Now, there is a trick at your disposal... you could do this:
```python
rr.log("random", rr.Points3D(positions, colors=colors, radii=radii))
rr.log_components("random", [rr.components.ColorBatch([255, 0, 0])], num_instances=2)
```

And now you'll end up with only red points, because you explicitly said that the data was 2 instances wide, and so the log function considers the `ColorBatch` to be a splat...

Of course we could change things so that logging 1 thing is always considered a splat, but then you have the opposite problem, which might or might not be better depending on the situation :shrug:.

[And this is why I don't like that splats are a logging-time rather than a query-time concern](https://github.com/rerun-io/rerun/issues/1893): the view should get to decide what to do with the data that it has as its disposal, and that behavior should be configurable through blueprints and through the UI.
This instance key business is pretty similar to e.g. configurable texture clamping modes in gfx APIs after all.